### PR TITLE
plo: device initialization using static constructors

### PR DESCRIPTION
**Pull request by HBuczynski**
This PR provides the following changes:
- add `_startc.c` to handle constructors and destructors
- add device interface to allow initialize specific driver based on platform settings

For example, registration of uart driver:

```
int uart_init(int dn, dev_handler_t *h)
{
...
}

__attribute__((constructor)) static void uart_regDev(void)
{
	devs_regDriver(DEV_UART_ID, uart_init);
}
```

If this mechanism is accepted I will implement it in the current drivers and will add changes it into this PR.

### CODE REVIEW

File: devices/devs.c
Comment (nalajcie) [inline devices/devs.c:22]: "which", but the correct solution is to zero the `bss` - which can easily be added to `_start.c` function as a C code now.
This would make the `active` variable unneeded (and other istances of initializing variables to `0` not needed also).

File: devices/devs.c
Comment (nalajcie) [inline devices/devs.c:33]: not sure if we want to have two-dimension array here, seems like a waste of space (I don't know about the use cases and if it's really important)

File: devices/devs.c
Comment (nalajcie) [inline devices/devs.c:77]: `devs_getHanlder` - typo

File: devices/devs.c
Comment (nalajcie) [inline devices/devs.c:70]: I assume this will be able to fail in the future and report it by return code?

File: devices/devs.c
Comment (nalajcie) [inline devices/devs.c:38]: `int dev` should probably be `unsigned int dev`, currently I can call `devs_regDriver(-3, NULL)` and overwrite BSS before the array :)

probably `int dn` should also be `unsigned`

File: devices/devs.c
Comment (nalajcie) [inline devices/devs.c:58]: would not be needed when we zero the BSS

File: devices/devs.h
Comment (nalajcie) [inline devices/devs.h:21]: is `config.h` needed for this header to parse?

File: devices/devs.h
Comment (nalajcie) [inline devices/devs.h:41]: I would add comment that this function can be called only before `devs_init`, preferably from device drivers constructors

**Review by nalajcie — COMMENTED** [review]
Some things need polishing, but I think it's the move in the right direction

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c]: Fixed

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c]: Yes, you have right. An oversight on my part.

File: devices/devs.h
Comment (HBuczynski) [reply devices/devs.h:21]: `config.h` contains platform specific typedefs. In this case, the `ssize_t` is located in config header.

File: devices/devs.h
Comment (HBuczynski) [reply devices/devs.h]: Ok, it can be helpful.

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c:70]: The report error will be printed by debug console which will not be depend on uart driver. 
I also wondering about plo behaviour when one of the driver will not be initialized correctly. Should it only report error or terminate as well?

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c]: The two-dimension array was introduced here to keep handlers of instances of all devices. For example, imxrt107x has 12 UART instances so that we need to be ready to keep them all.

File: devices/devs.c
Comment (nalajcie) [reply devices/devs.c:70]: In production environment I would try to boot the device nevertheless (who cares about uart5 anyway? :p). For debugging I think You will probably want to stop early to debug, but that can be done by placing asserts (either here or preferably at the point of failure - so in the device init function)

File: devices/devs.c
Comment (nalajcie) [reply devices/devs.c]: yeah, I understand WHY we need it (it's major/minor), just usually we will have 12 uarts and single ethernet / flash/ etc, so for most devices the other instances would be empty. I like the simplicity so If we're not afraid of the memory usage it's ok. We can always change the implementation to hashtable/linked list if that would provide to be problem on smaller targets.

File: _startc.c
Comment (kemonats) [inline _startc.c:32]: `_plo_bss` not  defined in `TARGET=armv7a9-zynq7000`

File: hal/armv7a/zynq7000/hal.c
Comment (kemonats) [inline hal/armv7a/zynq7000/hal.c:193]: If you are interested  I can provide optimized version of the `hal_memset()` function ( smaler size and uses fewer registers ).

Comment (HBuczynski) [reply hal/armv7a/zynq7000/hal.c:193]: This code comes from [phoenix-rtos-kernel/hal/armv7a](https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/master/hal/armv7a/string.h).
If you would like to optimize it, go ahead. However, I would like to ask you to do it in separate PR associated with phoenix-rtos-kernel. This thread hones in on devs module so I wouldn't like to discuss others, not related topics here.

Comment (kemonats) [reply hal/armv7a/zynq7000/hal.c:193]: I proposed this almost a year ago PR [#88](https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/88)

File: _startc.c
Comment (HBuczynski) [reply _startc.c]: Fixed

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c]: Ok

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c]: This initialization is already removed, BSS is zeroed.

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c]: BSS is zeroed. `active` variable has been removed.

File: hal/armv7a/zynq7000/hal.c
Comment (HBuczynski) [reply hal/armv7a/zynq7000/hal.c:192]: I will test it and add in a separate PR.

File: hal/armv7a/zynq7000/Makefile
Comment (nalajcie) [inline hal/armv7a/zynq7000/Makefile:14]: ok, now not so sure it will work, as the value of `__bss_start` will be `0`

File: hal/armv7a/zynq7000/config.h
Comment (nalajcie) [inline hal/armv7a/zynq7000/config.h:43]: maybe fix the linker symbols declaration here also?

**Review by nalajcie — CHANGES_REQUESTED** [review]
I think You have linker variable addresses and values mixed up

File: _startc.c
Comment (agkaminski) [reply _startc.c:16]: This symbol is just a symbol with no type. We have to give it some type to declare it in C. Declaring it as a function has an advantage that we can access this sybol without `&`. To get _end we need either:

```
extern char _end;

address = &_end;
```

or 
```

extern void _end(void);

address = _end;
```

Comment (nalajcie) [reply _startc.c:16]: I wrote the same in my next comment :). I believe directly accessing the variable address is more readable than defining it as a function and using the name as a function pointer (seems to me that the "more advanced knowledge of C" is needed to understand the function pointers syntax - for casual reader it seems we're using the `_end` value, not it's address).

The main issue with `_plo_bss` still exists (we're setting symbol **value** - not **address** - by `LDFLAGS` - and here by declaring it as a function - we use it's **address**). Actually, the existence of this bug is a proof that the previously used syntax is somewhat hard :P.

I'm fine with either approach, it just comes down to the readability.

Comment (nalajcie) [reply _startc.c:16]: sorry, my experiments on host target were invalid, indeed by using `--defsym` `LDFLAG` You would directly set the symbol **address**, the symbol will not be put in any section (will be marked as absent).

So the code works correctly (the only consideration is readability).

File: _startc.c
Comment (nalajcie) [reply _startc.c:34]: sorry, my experiments on host target were invalid, indeed by using --defsym LDFLAG You would directly set the symbol address, the symbol will not be put in any section (will be marked as absent).

So the code works correctly (the only consideration is readability).
Closing this as the code should work.

File: devices/devs.c
Comment (HBuczynski) [reply devices/devs.c]: Done

**Review by nalajcie — APPROVED** [review]
LGTM

File: cmd.c
Comment (kemonats) [inline cmd.c:889]: Since the maximum number of arguments is 4, it can declare 
```
	char args[5][LINESZ + 1]
```
save `5 * 81 ` bytes on the stack.

File: cmd.c
Comment (kemonats) [inline cmd.c:892]: The `cmd_skipblanks()` function is called at the begining of the `cmd_getnext()` function.

File: cmd.c
Comment (kemonats) [inline cmd.c:918]: If it will not been expanded:
```suggestion
	unsigned int pos = 0;

	cmd_skipblanks(s, &p, DEFAULT_BLANKS);
	s += p;

	if (*s) {
```

File: cmd.c
Comment (kemonats) [inline cmd.c:934]: The `cmd_skipblanks()` function is called at the begining of the `cmd_getnext()` function.

File: phfs.c
Comment (kemonats) [inline phfs.c:98]: Move `plostd_strlen(alias)` out of `for` loop, `alias` length does not change in a loop.

File: phfs.c
Comment (kemonats) [inline phfs.c:178]: Same as in `phfs_getHandlerId(const char *alias)`

File: phfs.c
Comment (kemonats) [inline phfs.c:207]: same as in `int phfs_regDev(const char *alias, unsigned int major, unsigned int minor, const char *prot)`

File: cmd.c
Comment (HBuczynski) [reply cmd.c]: You have right. Fixed

File: cmd.c
Comment (HBuczynski) [reply cmd.c]: You have right. Fixed

File: cmd.c
Comment (HBuczynski) [reply cmd.c]: Done

File: cmd.c
Comment (HBuczynski) [reply cmd.c]: You have right. Fixed

File: cmd.c
Comment (HBuczynski) [reply cmd.c]: Removed `cmd_skipblanks()`

File: phfs.c
Comment (HBuczynski) [reply phfs.c]: Thanks for noticing. Fixed

File: phfs.c
Comment (HBuczynski) [reply phfs.c]: Thanks for noticing. Fixed

File: phfs.c
Comment (HBuczynski) [reply phfs.c]: Done

File: phfs.c
Comment (HBuczynski) [reply phfs.c]: Ok, I agree with you, code is more readable and `plostd_strlen` is invoked only once.

**Comment by nalajcie** [comment]
generic comment regarding the development process:
 - I don't like that after I approved the PR there were additional commits added
 - let's not treat PRs as a long-lasting feature branch - let's just create the branch and make small approved PRs against it - and then rebase/merge the feature branch to master once it's ready
 - making PR grow in size makes it muuuch harder to review
 - (side note: if You want to mark PR as a draft - there is a dedicated button for that, don't use "DRAFT" label)
 - "draft" PRs mean they are not ready for review and I think nobody should review them. We already reviewed this PR so it should not be marked as draft.

Because of the above I'm withdrawing my PR approval and don't plan to review this PR anymore.

**Review by nalajcie — COMMENTED** [review]
PR approval removed.

**Comment by HBuczynski** [comment]
Due to the fact that this PR has grown to big size, it will be divided into the smaller ones to make changes more readable during review.
Comments discussed in this section are moved to other PRs.

**Pull request closed without merging**
